### ISSUE TITLE
healthcheck.sh script timeout

### ISSUE THREAD

**Issue by kfirfer** [bitnami/bitnami-docker-pgpool#35]
Hello

I getting sometimes timeouts from the "healthckeck.sh" script (rarely but happens)

Im using docker image: "bitnami/pgpool:4.1.3-debian-10-r21"

Setup:
* Running on kubernetes cluster
* Helm chart: **postgresql-ha version: 3.6.1**
* 3 replicas of postgres instances, and 1 pgpool instance

**Comment by FraPazGal** [comment]
Hello @kfirfer,

I am not sure if I have understood correctly, but you are using our [postgres-ha chart](https://github.com/bitnami/charts/tree/master/bitnami/postgresql-ha), is that right?

I have deployed said chart using your setup but I can't reproduce the issue. In one of the deployments the pgpool pod was slower than normal and its liveness probe (exec the `healthcheck.sh` script) failed once, but it then proceeded as expected. Do the timeouts for the probe occur at deployment time or do they pop-up from time to time when the chart has been running for a while?

Could you please provide info on how you are installing the chart and what the output of `kubectl describe` for your pgpool pod is?

**Comment by kfirfer** [comment]
Hi

Yes Im using the postgres-ha chart
And im getting timeouts for the probe time to time
It is rarely happens 

described of pgpool:

```
vostro@dev101:~/code/zigi/zigi-server-java$ kubectl describe po postgresql-ha-pgpool-67c84549d5-d4ktk                       
Name:         postgresql-ha-pgpool-67c84549d5-d4ktk
Namespace:    harbor
Priority:     0
Node:         nuc03/192.168.200.103
Start Time:   Fri, 18 Sep 2020 03:13:26 +0300
Labels:       app.kubernetes.io/component=pgpool
              app.kubernetes.io/instance=postgresql-ha
              app.kubernetes.io/managed-by=Helm
              app.kubernetes.io/name=postgresql-ha
              helm.sh/chart=postgresql-ha-3.6.1
              pod-template-hash=67c84549d5
Annotations:  cni.projectcalico.org/podIP: 172.16.206.205/32
              cni.projectcalico.org/podIPs: 172.16.206.205/32
              kubernetes.io/psp: restricted
Status:       Running
IP:           172.16.206.205
IPs:
  IP:           172.16.206.205
Controlled By:  ReplicaSet/postgresql-ha-pgpool-67c84549d5
Containers:
  pgpool:
    Container ID:   docker://c4c248555c0a1c8c4e4f308ed964038dba69cce3ad80305af4912df25a7202a0
    Image:          docker.io/bitnami/pgpool:4.1.3-debian-10-r21
    Image ID:       docker-pullable://bitnami/pgpool@sha256:e8a198007ca3abe993f2af1d241b65b7ec31c9483548b1bdb6c5201b121f1e99
    Port:           5432/TCP
    Host Port:      0/TCP
    State:          Running
      Started:      Fri, 18 Sep 2020 03:13:26 +0300
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     1
      memory:  1Gi
    Requests:
      cpu:      50m
      memory:   150Mi
    Liveness:   exec [/opt/bitnami/scripts/pgpool/healthcheck.sh] delay=30s timeout=10s period=10s #success=1 #failure=5
    Readiness:  exec [bash -ec PGPASSWORD=${PGPOOL_POSTGRES_PASSWORD} psql -U "postgres" -d "postgres" -h /opt/bitnami/pgpool/tmp -tA -c "SELECT 1" >/dev/null] delay=5s timeout=5s period=5s #success=1 #failure=5
    Environment:
      BITNAMI_DEBUG:                 false
      PGPOOL_BACKEND_NODES:          0:postgresql-ha-postgresql-0.postgresql-ha-postgresql-headless.harbor.svc.cluster.local:5432,1:postgresql-ha-postgresql-1.postgresql-ha-postgresql-headless.harbor.svc.cluster.local:5432,2:postgresql-ha-postgresql-2.postgresql-ha-postgresql-headless.harbor.svc.cluster.local:5432,
      PGPOOL_SR_CHECK_USER:          repmgr
      PGPOOL_SR_CHECK_PASSWORD:      <set to the key 'repmgr-password' in secret 'postgresql-ha-postgresql'>  Optional: false
      PGPOOL_ENABLE_LOAD_BALANCING:  yes
      PGPOOL_ENABLE_LDAP:            no
      PGPOOL_POSTGRES_USERNAME:      postgres
      PGPOOL_POSTGRES_PASSWORD:      <set to the key 'postgresql-password' in secret 'postgresql-ha-postgresql'>  Optional: false
      PGPOOL_ADMIN_USERNAME:         admin
      PGPOOL_ADMIN_PASSWORD:         <set to the key 'admin-password' in secret 'postgresql-ha-pgpool'>  Optional: false
      PGPOOL_ENABLE_TLS:             no
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-2rnhr (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             True 
  ContainersReady   True 
  PodScheduled      True 
Volumes:
  default-token-2rnhr:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-2rnhr
    Optional:    false
QoS Class:       Burstable
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute op=Exists for 15s
                 node.kubernetes.io/unreachable:NoExecute op=Exists for 15s
Events:          <none>

```

**Comment by FraPazGal** [comment]
Hi @kfirfer,

Does the liveness probe timeout develop into a crash or it just timeouts once and then the pod keeps running correctly?

I'll leave my testing cluster running, as I have yet to encounter any timeouts. Does it happens when pgpool is under heavy load or does it also occurs when idle? If it is posible, could you paste the logs of the pod when there is a timeout? Thanks in advance.

**Comment by kfirfer** [comment]
I changed the the db to the postgresql chart instead postgresql-ha , the ha chart seems less stable from the regular chart:
 
https://github.com/bitnami/bitnami-docker-postgresql-repmgr/issues/37

when this issue resolve I will go back to the ha chart

**Comment by juan131** [comment]
Hi @vikram-bitnami 

I just answered you to the issue with PostgreSQL HA. Let's go back to this conversation once it's solved.

**Comment by dominik-bln** [comment]
To give one more "data point": We are also seeing liveness probes errors on a pgpool deployed via the HA chart about once a day, often leading to a restart of the Pod. This only happens on the production instance, staging runs fine.

Resource usage seems ok though, but it coincides with a small increase in memory consumption right before the problems:

<img width="1466" alt="Screenshot 2020-11-27 at 10 11 11" src="https://user-images.githubusercontent.com/855119/100431791-ee883c00-3098-11eb-8582-5026dc0825ab.png">

<img width="1221" alt="Screenshot 2020-11-27 at 10 07 59" src="https://user-images.githubusercontent.com/855119/100431663-c8629c00-3098-11eb-8d06-28697e44f242.png">

**Comment by javsalgar** [comment]
Hi,

That's interesting. Thanks for the information! I don't know if it may make sense to relax the liveness probe conditions for the time being. It is not clear if that increase in memory is expected or it has to do with a bug in the configuration.

**Comment by dominik-horb-umg** [comment]
We relaxed the liveness probe settings from the Chart to `Liveness:   exec [/opt/bitnami/scripts/pgpool/healthcheck.sh] delay=30s timeout=20s period=20s #success=1 #failure=5` yesterday morning and haven't seen a restart since. I'll report back if anything comes up.

**Comment by javsalgar** [comment]
Good to know! Then we should consider using these relaxed values as the default.

**Comment by dominik-bln** [comment]
We are still seeing occasional timeouts of the script, but it doesn't look like it's causing restarts anymore and as far as I can tell the application using the service is available the whole time:

<img width="1083" alt="Screenshot 2020-12-04 at 10 10 13" src="https://user-images.githubusercontent.com/855119/101144679-fad04400-3618-11eb-8f70-7ff045089b8f.png">

**Comment by javsalgar** [comment]
Great! Would you like to create a PR with the new relaxed settings?

**Comment by dominik-bln** [comment]
I'd like to observe a few days more if that is fine with you. Just want to make sure that I'm not masking any real timeouts for the application with these new settings. I will open an MR on the `postgres-ha` repo in a few days then.

**Comment by javsalgar** [comment]
Ok! Thanks a lot!

**Comment by dominik-bln** [comment]
Sadly I have to report that the problem really seems to have moved to the application now, i. e. we are seeing occasional readiness probe errors of the Django App that is using the connection. I cannot fully rule out some other problem, but it coincides well enough with the pgool liveness probe changes and it occurs at about the same frequency.

This would indicate to me that the the liveness probe timeout is showing a genuine problem. I'll try to find some time to look at this further, but not sure when I'll get to this.

**Comment by marcosbc** [comment]
Thanks for sharing. I'm going to update our internal task with this information.

Would you happen to be able to share some steps to reproduce this issue in a consistent manner? (with sample data and so, if needed)

**Comment by souvik-deepsource** [comment]
@juan131 Any update on this. I just met with the same issue. I'm using the PostgreSQL-HA chart and using `docker.io/bitnami/pgpool:4.1.2-debian-10-r23` as the Pgpool image. The pgpool container had abruptly restarted without any logs.

**Comment by marcosbc** [comment]
Unfortunately no update @souvik-deepsource. Does it keep restarting (i.e. CrashLoopBackOff) or was it a one-time thing?

Also, do you happen to have some steps to reproduce this issue in a consistent manner from our side?

**Comment by s-kunze** [comment]
Hi all,

We also observed the described situation, but we could see different behaviors with the number of applications are using the database (we also use the bitnami postgresql-ha chart) along with the number of connections to the database those applications are creating.

For applications that are using fewer connections, everything worked fine.

But we have some applications which used quite more connections. For them, we could also see that pgpool was restarting because of the health check. 

After many tries, we solved the issue by updating the value `numInitChildren` to `64` instead of `32` (default). When I get the documentation correctly, this is also the maximum number of parallel connections to the Postgresql Database?

**Comment by marcosbc** [comment]
@s-kunze The maximum number of parallel connections is proportional to that value, but also max_pool, see: https://www.pgpool.net/mediawiki/index.php/Relationship_between_max_pool,_num_init_children,_and_max_connections

**Comment by foarsitter** [comment]
Thanks for pointing me to the wiki @marcosbc. I'm now running for more than three day's without a healthcheck timeout / restart. Before this I experienced daily restarts. 

I'm running postgres now with `max_connections=200` and the two pgpool instances with `max_pool=15` (default) and `num_init_children=64` and it is running smooth.

**Comment by oliverwiegers** [comment]
Hello together,

I might have just found the issue. 
I had livenessProbe timeouts in the past and those vanished after making some changes alongside to an update. So i thought the update fixed an issue.
Today after an update the livenessProbe timeouts occurred again.
I'm using the bitnami/postgresql-ha helm chart. 

After making sure the helm chart had no configuration issues I went on to investigate the livenessProbe which executes the script found in the conainter at: `/opt/bitnami/scripts/pgpool/healthcheck.sh`:

```
#!/bin/bash
#
# Bitnami Pgpool healthcheck

# shellcheck disable=SC1091

set -o errexit
set -o nounset
set -o pipefail
# set -o xtrace # Uncomment this line for debugging purpose

# Load libraries
. /opt/bitnami/scripts/libpgpool.sh

# Load Pgpool env. variables
eval "$(pgpool_env)"

pgpool_healthcheck
```

The part that is taking long is the evaluation of `pgpool_env`:

```
❯ k exec -it pod/postgresql-ha-pgpool-59895cccc4-flml6 -- /bin/bash
I have no name!@postgresql-ha-pgpool-59895cccc4-flml6:/$ . /opt/bitnami/scripts/libpgpool.sh
I have no name!@postgresql-ha-pgpool-59895cccc4-flml6:/$ time eval "$(pgpool_env)"

real    3m11.602s
user    0m0.017s
sys     0m0.012s
```

`pgpool_env` is a function defined in `/opt/bitnami/scripts/libpgpool.sh` which has many `export` statements based on the environment variables passed to the container to setup the PostgreSQL envrionment.
Based on the amount of configuration values and the length of their content the evaluation process takes time.
In my case I have configured a bunch of `PGPOOL_POSTGRES_CUSTOM_USERS` and `PGPOOL_POSTGRES_CUSTOM_PASSWORDS`.

Of course the execution time not only depends on configuration values but on pod resources as well. If run by kubernetes.

I hope this might help.
Have a nice weekend

**Comment by phntom** [comment]
i'm also experiencing this issue, i've tried restoring from backup on a clean install with postgresql-ha 8.1.2 (11.14.0) and i experienced both very slow liveliness checks and about 3 minutes for the first logs to arrive.
both postgresql pods are fine, this only affects pgpool.
disabling liveliness checks is the only workaround at the moment.
also i've tried upgrading with it enabled an a high timeout and the pod came up almost immediately so i'm not sure what's the issue but i can replicate it here.

anyone knows of an older version where this doesn't happen?

**Comment by juan131** [comment]
Hopefully these changes can improve the Pgpool restarts: https://github.com/bitnami/charts/pull/8398

**Comment by abhi1693** [comment]
For me adding this to the config, finally dropped the number of restarts from a few hundred per hour to less than 3.

```
pgpool:
    configuration: |-
      health_check_timeout=30
      health_check_period=10
      health_check_max_retries=10
      failover_on_backend_error='off'
      health_check_retry_delay=5
    livenessProbe:
      timeoutSeconds: 15
    readinessProbe:
      initialDelaySeconds: 30
      periodSeconds: 10
      timeoutSeconds: 15
```

**Comment by gondaz** [comment]
Hello everyone!
Just a small update from me.
None of relaxing the script or increasing init connections helped me, I always encountered multiple restarts (5 to 10 per hour) of pod during a liveness probe that could not be executed successfully.
Without a liveness probe everything works perfect, no restarts at all, so I started to investigate what could be wrong during the liveness probe script execution.
I started a looped healthcheck script from my terminal in order to catch something wrong during execution:
`while :; do kubectl -n some-namespace exec -it $(kubectl get pods -n some-namespace -l app=pgpool -o jsonpath="{.items[0].metadata.name}") -- bash -x /opt/bitnami/scripts/pgpool/healthcheck.sh; sleep 10; done`
After a while, I noticed, that there was a very slow step when this script was executing:
```
--------------------------------
+++ head -c 20 /dev/random
+++ base64
--------------------------------
```
Every time head command execution could take up to 20s to complete, so I thought this is the reason why kubelet could not execute the healthcheck script properly and just exited abruptly (timeout exception).  Then I tried to execute that command directly at the k8s worker node where the pgpool pod was running, and result was the same. It seems there was a lack of entropy so any randomizing operations were very slow as well.
Then I installed haveged package at my k8s worker nodes:
`sudo apt install haveged`
and the issue is gone, as of now the healthcheck script is not stucking anymore and therefore no restarts for pgpool.
My liveness probe settings are following:
```
          livenessProbe:
            exec:
              command:
                - /opt/bitnami/scripts/pgpool/healthcheck.sh
            initialDelaySeconds: 40
            periodSeconds: 30
            timeoutSeconds: 5
            successThreshold: 1
            failureThreshold: 2
```
Hopefully, these steps will help someone.
